### PR TITLE
Chat badges

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Adds in chat badges.

This system is like the previous chat PR, expect the DB changes are simplified out. Badges.json loads the icon_state of each admin rank, Mentor and Donator rank.

```
On world initialization, badge ranks are loaded from the json file with format:
{
    "Admin Observer": "adminobserver",
    "Moderator": "moderator",
...
}
```
The value on the left is the name of the admin rank, and the value on the right being the icon_state of the badge. The rank name is matched to the name in the permissions panel.
Additionally there are 2 special rank names in this file: Mentor and Donator.
Mentor matches the badges to people with a mentor_datum and Donator matches the badge to people who are in the patrons list.

The badges a client have are cached and only loaded when they change badge status (through deadmin, permission panel change, losing mentor etc.) to improve performance (List checking against the patron list all the time will be slow).

Procs:
/proc/load_badge_ranks() - Loads the badges json and resets the badge status of all clients.
/client/proc/get_badges() - Gets the badges attached to a client and caches them. Will return the clients cached_badges var if it is a list, if not will get all the badges attached to that client and then store them in the cached_badges var.
/client/proc/reset_badges() - Sets the cached_badges of a client to null which will in turn cause them to be reloaded when get_badges is called next. This needs to be called when the badge state of a client changes through deadminning or other means.
/proc/badge_parse(badges) - Converts a list of badge icon_states to a parsed thing.

image

image

image

image

Deadmin removes the admin rank, stealthmin removes 

## Why It's Good For The Game

Adds in an admin identifier, something for donators and a badge for mentors.

## Changelog
:cl:
add: Adds in chat badges for admin ranks, mentors and donators.
/:cl:

### CODE REVIEW

**Comment by PowerfulBacon** [comment]
Adds a space between the badges to look like original

image

**Comment by ItsTisme** [comment]
Looks nice, deffintly fun addition to flex with

**Comment by aidan-bowman** [comment]
could use a little bit more variety (8 banhammers eef) but maybe it's just me

**Comment by Mat05usz** [comment]
> could use a little bit more variety (8 banhammers eef) but maybe it's just me

3 of those hammers are never used, 1 very rarely so I think it's fine.

**Comment by OceanFish1** [comment]
we had coder role ingame?

**Comment by PowerfulBacon** [comment]
Only host, headmin, seniormind, admin, trialmin, coder, headdev mentor and donator I have seen in game, the others were just in the files so I added them anyway.

**Comment by Mat05usz** [comment]
> we had coder role ingame?

I remember Franc having a Coder role for a long time on Bee

File: code/modules/badges/badges.dm [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+//Loads the badge ranks
+/proc/load_badge_ranks()
+	//No badges
+	if(!CONFIG_GET(flag/badges))
+		return
+	//Load and parse data
+	GLOB.badge_data = json_decode(rustg_file_read("[global.config.directory]/badges.json"))
+	//Associate badges with admin ranks
+	for(var/datum/admin_rank/rank in GLOB.admin_ranks)
```
Comment (Archanial) [inline code/modules/badges/badges.dm]:
```suggestion
	for(var/datum/admin_rank/rank as() in GLOB.admin_ranks)
```
there are only ranks in this list right?

File: code/modules/badges/badges.dm [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		rank.badge_icon = GLOB.badge_data[rank.name]
+	//Yay
+	log_game("[LAZYLEN(GLOB.badge_data)] badges loaded successfully.")
+	//Reset everyones badges so they get reloaded.
+	for(var/client/C in GLOB.clients)
```
Comment (Archanial) [inline code/modules/badges/badges.dm]:
```suggestion
	for(var/client/C as() in GLOB.clients)
```
same here I guess?

File: code/modules/badges/badges.dm [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		C.reset_badges()
+
+//Gets the badges attached to a client.
+/client/proc/get_badges()
+	//No badges
+	if(!CONFIG_GET(flag/badges))
+		return
+	//Send cached badges
+	if(islist(cached_badges))
```
Comment (Archanial) [inline code/modules/badges/badges.dm:44]:
```suggestion
	if(LAZYLEN(cached_badges))
```

**Review by Archanial — COMMENTED** [review]
Not sure how you want badges like coder or other not tied to ranks to appear

**Comment by ike709** [comment]
I think headmin and seniormin colors should swap, so it gets progressively more of a dark red the more screwed the tider is when they see it.

**Comment by PowerfulBacon** [comment]
> 
> 
> I think headmin and seniormin colors should swap, so it gets progressively more of a dark red the more screwed the tider is when they see it.

Its how it is in discord

image

File: code/modules/badges/badges.dm (hunk 3 above)
Comment (PowerfulBacon) [reply code/modules/badges/badges.dm:44]: I want to check if its a list and not null, not if it has items in it

**Comment by PowerfulBacon** [comment]
> 
> 
> Not sure how you want badges like coder or other not tied to ranks to appear

Coder is a rank

**Comment by Archanial** [comment]
> 
> 
> > Not sure how you want badges like coder or other not tied to ranks to appear
> 
> Coder is a rank

Maintainer is coder in game but you have badges for both.

**Comment by ike709** [comment]
Fair enough.

File: code/modules/badges/badges.dm (hunk 3 above)
Comment (Archanial) [reply code/modules/badges/badges.dm:44]: Then why not use `isnull`?

**Comment by DatBoiTim** [comment]
How many badges are possible to display at once?

File: code/modules/badges/badges.dm (hunk 3 above)
Comment (MistakeNot4892) [reply code/modules/badges/badges.dm:44]: Because isnull doesn't check if it's a list.

**Comment by PowerfulBacon** [comment]
> How many badges are possible to display at once?

1 for your admin rank, 1 for your mentor and 1 for donator

File: code/modules/admin/admin_verbs.dm [hunk 4]
```diff
@@ -296,6 +296,7 @@ GLOBAL_PROTECT(admin_verbs_hideable)
 				add_verb(/client/proc/play_web_sound)
 		if(rights & R_SPAWN)
 			add_verb(GLOB.admin_verbs_spawn)
+		reset_badges()
```
Comment (qwertyquerty) [inline code/modules/admin/admin_verbs.dm:299]: Should this really be in this proc?

File: code/modules/admin/admin_verbs.dm [hunk 5]
```diff
@@ -321,6 +322,7 @@ GLOBAL_PROTECT(admin_verbs_hideable)
 		/client/proc/fix_say
 		)
 	remove_verb(verb_list)
```
Comment (qwertyquerty) [inline code/modules/admin/admin_verbs.dm:324]: Same here

File: code/modules/admin/admin_verbs.dm [hunk 6]
```diff
@@ -321,6 +322,7 @@ GLOBAL_PROTECT(admin_verbs_hideable)
...
+	reset_badges()
```
Comment (qwertyquerty) [inline code/modules/admin/admin_verbs.dm:325]: Same here

File: code/modules/admin/admin_verbs.dm [hunk 7]
```diff
@@ -490,6 +492,7 @@ GLOBAL_PROTECT(admin_verbs_hideable)
 	if(holder)
 		if(holder.fakekey)
 			holder.fakekey = null
+			reset_badges()
```
Comment (qwertyquerty) [inline code/modules/admin/admin_verbs.dm:495]: Ditto

File: code/modules/admin/admin_verbs.dm [hunk 8]
```diff
@@ -500,6 +503,7 @@ GLOBAL_PROTECT(admin_verbs_hideable)
 			if(!new_key)
 				return
 			holder.fakekey = new_key
+			reset_badges()
```
Comment (qwertyquerty) [inline code/modules/admin/admin_verbs.dm:506]: Ditto

File: code/modules/badges/badges.dm [hunk 9]
```diff
@@ -0,0 +1,88 @@
+/*
+
+Loads all badges from the badge_ranks.txt file.
```
Comment (qwertyquerty) [inline code/modules/badges/badges.dm]: outdated docs

File: code/modules/badges/badges.dm [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+Icon - The name of the icon state in the badges.dmi file
+
+Special ranks:
+- Donator: Matches clients in the donator list
+- Mentor: Matches clients in the mentor list
+
+*/
+
+//Global list of badges
+GLOBAL_LIST_EMPTY(badge_data)
```
Comment (qwertyquerty) [inline code/modules/badges/badges.dm:17]: C tier name

**Comment by qwertyquerty** [comment]
Do we really think badges are expensive enough to justify caching them? It would save the trouble of knowing every single instance in which they need to be reset.

**Comment by utilbird** [comment]
\>no twitch mod badge
\>no reddit gold

Coder and maintainer are very hard to distinguish, they could probably use some tweaking

**Review by utilbird — COMMENTED** [review]
Badges should probably be grabbed and stored for each client on connection so badges aren't fetched every time they send an OOC messages

**Comment by PowerfulBacon** [comment]
> 
> 
> Badges should probably be grabbed and stored for each client on connection so badges aren't fetched every time they send an OOC messages

Thats exactly how it works, except they are grabbed the first time they send an OOC message and cached for every message afterwards. It works the same was as if they were cached on connection, except if a client joins and never does something that requires a badge to be fetched then no badges will be cached to the client.

**Comment by PowerfulBacon** [comment]
> 
> 
> Do we really think badges are expensive enough to justify caching them? It would save the trouble of knowing every single instance in which they need to be reset.

The IS_PATRON define seems pretty intensive to be running each OOC message since it runs a find against a potentially pretty large list.

**Comment by EgorDinamit** [comment]
> The IS_PATRON define seems pretty intensive to be running each OOC message since it runs a find against a potentially pretty large list.

Doesn't it just run through a global list of patrons that is loaded on initialization in search of a ckey?
Shouldn't affect performance that much, imo.

**Comment by aidan-bowman** [comment]
couldn't you just make it run once when the player joins, then attach a variable to the player and simply check that variable when someone chats?

**Comment by EgorDinamit** [comment]
> couldn't you just make it run once when the player joins, then attach a variable to the player and simply check that variable when someone chats?

So, a "holder", but for patrons.
image
Shouldn't be hard by the way.

**Comment by utilbird** [comment]
> Thats exactly how it works, except they are grabbed the first time they send an OOC message and cached for every message afterwards. It works the same was as if they were cached on connection, except if a client joins and never does something that requires a badge to be fetched then no badges will be cached to the client.

Oh! Alright, my bad 😅

**Comment by PowerfulBacon** [comment]
> 
> 
> couldn't you just make it run once when the player joins, then attach a variable to the player and simply check that variable when someone chats?

Thats what it does

**Comment by EgorDinamit** [comment]
> > couldn't you just make it run once when the player joins, then attach a variable to the player and simply check that variable when someone chats?
> 
> Thats what it does

I think they are referring to the IS_PATRON define, which does the ckey lookup in a list.

**Comment by ike709** [comment]
Please address the reviews, otherwise this PR is good to go.

File: code/modules/admin/admin_verbs.dm (hunk 4 above)
Comment (PowerfulBacon) [reply code/modules/admin/admin_verbs.dm:299]: Yes, it resets the badge cache allowing the badges to be reloaded

File: code/modules/admin/admin_verbs.dm (hunk 6 above)
Comment (PowerfulBacon) [reply code/modules/admin/admin_verbs.dm:325]: Same as above with all of them

File: code/modules/badges/badges.dm (hunk 10 above)
Comment (PowerfulBacon) [reply code/modules/badges/badges.dm:17]: It's a list of the data associated with badges, what else can I put for it?

File: code/modules/badges/badges.dm (hunk 9 above)
Comment (PowerfulBacon) [inline code/modules/badges/badges.dm]:
```suggestion
Loads all badges from the badge.json file.
```

**Comment by PowerfulBacon** [comment]
done

**Pull request merged**